src/ring_theory/integral_closure.lean

```diff
@@ -0,0 +1,280 @@
+  rw [finset.mem_range, nat.lt_succ_iff], refine le_of_not_lt (λ hk, _),
+  rw [degree_le_iff_coeff_zero] at this,
+  rw [finsupp.mem_support_iff] at hkq, apply hkq, apply this,
+  exact lt_of_le_of_lt degree_le_nat_degree (with_bot.coe_lt_coe.2 hk)
+end
+
+theorem fg_adjoin_of_finite {s : set A} (hfs : s.finite)
+  (his : ∀ x ∈ s, is_integral R x) : (algebra.adjoin R s : submodule R A).fg :=
+set.finite.induction_on hfs (λ _, ⟨finset.singleton 1, le_antisymm
+  (span_le.2 $ set.singleton_subset_iff.2 $ is_submonoid.one_mem _)
+  (show ring.closure _ ⊆ _, by rw set.union_empty; exact
+    set.subset.trans (ring.closure_subset (set.subset.refl _))
+    (λ y ⟨x, hx⟩, hx ▸ mul_one (algebra_map A x) ▸ algebra.smul_def x (1:A) ▸ (mem_coe _).2
+      (submodule.smul_mem _ x $ subset_span $ or.inl rfl)))⟩)
+(λ a s has hs ih his, by rw [← set.union_singleton, algebra.adjoin_union_coe_submodule]; exact
+  fg_mul _ _ (ih $ λ i hi, his i $ set.mem_insert_of_mem a hi)
+    (fg_adjoin_singleton_of_integral _ $ his a $ set.mem_insert a s)) his
+
+theorem is_integral_of_noetherian' (H : is_noetherian R A) (x : A) :
+  is_integral R x :=
+begin
+  let leval : @linear_map R (polynomial R) A _ _ _ _ _ := (aeval R A x).to_linear_map,
+  let D : ℕ → submodule R A := λ n, (degree_le R n).map leval,
+  let M := well_founded.min (is_noetherian_iff_well_founded.1 H)
+    (set.range D) (set.ne_empty_of_mem ⟨0, rfl⟩),
+  have HM : M ∈ set.range D := well_founded.min_mem _ _ _,
+  cases HM with N HN,
+  have HM : ¬M < D (N+1) := well_founded.not_lt_min
+    (is_noetherian_iff_well_founded.1 H) (set.range D) _ ⟨N+1, rfl⟩,
+  rw ← HN at HM,
+  have HN2 : D (N+1) ≤ D N := classical.by_contradiction (λ H, HM
+    (lt_of_le_not_le (map_mono (degree_le_mono
+      (with_bot.coe_le_coe.2 (nat.le_succ N)))) H)),
+  have HN3 : leval (X^(N+1)) ∈ D N,
+  { exact HN2 (mem_map_of_mem (mem_degree_le.2 (degree_X_pow_le _))) },
+  rcases HN3 with ⟨p, hdp, hpe⟩,
+  refine ⟨X^(N+1) - p, monic_X_pow_sub (mem_degree_le.1 hdp), _⟩,
+  show leval (X ^ (N + 1) - p) = 0,
+  rw [linear_map.map_sub, hpe, sub_self]
+end
+
+theorem is_integral_of_noetherian (M : subalgebra R A)
+  (H : is_noetherian R (M : submodule R A)) (x : A) (hx : x ∈ M) :
+  is_integral R x :=
+begin
+  letI : algebra R M := M.algebra,
+  letI : comm_ring M := M.comm_ring R A,
+  suffices : is_integral R (⟨x, hx⟩ : M),
+  { rcases this with ⟨p, hpm, hpx⟩,
+    replace hpx := congr_arg subtype.val hpx,
+    refine ⟨p, hpm, eq.trans _ hpx⟩,
+    simp only [aeval_def, eval₂, finsupp.sum],
+    rw ← finset.sum_hom subtype.val,
+    { refine finset.sum_congr rfl (λ n hn, _),
+      change _ = _ * _,
+      rw is_semiring_hom.map_pow subtype.val, refl,
+      split; intros; refl },
+    refine { map_add := _, map_zero := _ }; intros; refl },
+  refine is_integral_of_noetherian' H ⟨x, hx⟩
+end
+
+set_option class.instance_max_depth 100
+theorem is_integral_of_mem_of_fg (M : subalgebra R A)
```

Comment:
That's a pretty long proof... are you sure you can't factor out helper lemmas? The informal proof is only 2 lines.

Reply:
What is the 2-line informal proof?

Reply:
Sorry, I take that back. I was confused.